man/utmpset.8
.TH utmpset 8
.SH NAME
utmpset \- logout a line from utmp and wtmp file
.SH SYNOPSIS
.B utmpset
[
.B \-w
]
.I line
.SH DESCRIPTION
The
.B utmpset
program modifies the user accounting database
.BR utmp (5)
and optionally
.BR wtmp (5)
to indicate that the user on the terminal
.I line
has logged out.
.P
Ordinary
.BR init (8)
processes handle utmp file records for local login accounting.
The
.BR runit (8)
program doesn't include code to update the utmp file, the
.BR getty (8)
processes are handled the same as all other services.
.P
To enable local login accounting, add
.B utmpset
to the
.BR getty (8)
.I finish
scripts, e.g.:
.P
 $ cat /service/getty-5/finish
 #!/bin/sh
 exec utmpset \-w tty5
 $
.SH OPTIONS
.TP
.B \-w
wtmp. Additionally to the utmp file, write an empty record for
.I line
to the wtmp file.
.SH EXIT CODES
.B utmpset
returns 111 on error, 1 on wrong usage, 0 in all other cases.
.SH SEE ALSO
sv(8),
runsv(8),
runit(8),
runit-init(8)
runsvdir(8),
runsvchdir(8),
chpst(8),
svlogd(8),
getty(8)
.P
https://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape@smarden.org>


package/CHANGES
  * doc/index.html: replace broken link to sysvinit.
  * doc/runscripts.html: remove broken links and reference to mailing list.
  * doc/replaceinit.html: switch http links to https.
  * alloc.c, alloc_re.c, buffer_0.c, byte.h, byte_chr.c, byte_copy.c,
    byte_cr.c, byte_diff.c, byte_rchr.c, select.h2, wait.h, wait_nohang.c,
    wait_pid.c: C23: swap ANSI for remaining K&R-style declarations
    (github/andy-bower).
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, pathexec_env.c, runit.c, runsv.c, runsvdir.c, select.h1,
    select.h2, sig.c, sig.h, sig_catch.c, svlogd.c, svlogd.dist, wait.h:
    fix build with gcc15: in C23 f() means f(void), consequently signal
    handler functions now are f(int), alloc_free() parameter needs cast,
    buffer_unixwrite() removes "const" from parameter.
  * doc/index.html: repair or remove broken links in list of distributions
    using runit, add Debian GNU/Hurd.
  * doc/*, man/*: switch all cr.yp.to http links to https.
  * doc/faq.html: fix link to LSB core specification.
  * doc/index.html: remove section "Related links".
  * doc/index.html, faq.html: update links to supervision mailing list and
    list archives.
  * doc/benefits.html: remove references to Debian packages.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * check-diff: check that files exist; quote properly.
  * utmpset.c: check length of line argument to match struct utmp; properly
    handle line argument with max length (thx github/Skyb0rg007).
  * doc/index.html: add reference to repository on github.com.
  * doc/faq.html: typo.
  * doc/*.html: !DOCTYPE.
  * doc/index.html: add antiX Linux to the list of distributions (thx C.
    Jarry).
  * hasshsgr.h?, warn-shsgr: remove; obsolete.
  * runit.c: don't leak SIGCONT SIG_IGN into all child processes (thx Leah
    Neukirchen).
  * doc/index.html: fix link to Void Linux (thx R.G.).

2.2.0
Fri, 27 Sep 2024 00:38:25 +0000
  * chpst.dist, sv.dist, svlogd.dist, utmpset.dist: adapt.
  * chpst.c, man/chpst.8: add -C option to change working directory (thx
    Andrew J. Hesford, void linux).
  * doc/install.html: add link to sha256sum.
  * doc/install.html, doc/replaceinit.html: change mail address.
  * package/TODO: remove.
  * package/README,COPYING: Copyright date.
  * chpst.c: fix compilation warning(s): -Wunsequenced (thx Dmitry
    Bogatov, Z. Liu).
  * svlogd.c: Initialize all fields of sockaddr_in (thx Érico Rolim).
  * sv.c: add sleep_microseconds to encapsulate (u|nano)sleep (thx Z.
    Liu).
  * chpst.c, chpst.dist, man/chpst.8: add options -t and -V to chpst man
    page and usage (thx Lorenzo Puliti, Z. Liu).
  * man/runsv.8: log service's ./finish script is actually not invoked
    (thx Jamie Heilman, Dmitry Bogatov, Z. Liu).
  * man/runsv.8: clarify how signal override works with control/d and x
    (thx Andras Korn, Lorenzo Puliti, Z. Liu).
  * runit.c, runit.h, man/runit.8: support /etc/runit/nosync file to make
    sync on shutdown/reboot optional (thx Andras Korn).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c: remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * chpst.c: replace prot_uid() with setuid().
  * utmpset.c: don't pass int32_t to time() (thx Lorenzo Beretta).
  * chpst.c, pathexec.h, pathexec_env.c, pathexec_run.c, runsvdir.c,
    sgetopt.c, sgetopt.h, subgetopt.c, subgetopt.h, sv.c, svlogd.c,
    utmpset.c: fix compilation warning(s): -Wincompatible-pointer-types-
    discards-qualifiers (thx Z. Liu, Debian, and others).
  * x86cpuid.c: fix error: call to undeclared library function '...'; ISO
    C99 and later do not support implicit function declarations
    [-Wimplicit-function-declaration] (thx Z. Liu, Debian, and others).
  * chpst.c, pathexec_run.c, prot.c, seek_set.c: fix error: call to
    undeclared function '...'; ISO C99 and later do not support implicit
    function declarations [-Wimplicit-function-declaration] (thx Z. Liu).
  * tryflock.c, trysgact.c, trysgprm.c, trywaitp.c: fix error: type
    specifier missing, defaults to 'int'; ISO C99 and later do not support
    implicit int [-Wimplicit-int] (thx Z. Liu and others).
  * chkshsgr.c, trycpp.c, trypoll.c, tryshsgr.c, tryulong64.c: fix error:
    call to undeclared function '...'; ISO C99 and later do not support
    implicit function declarations [-Wimplicit-function-declaration] (thx
    Z. Liu and others).

2.1.2
Sun, 10 Aug 2014 18:01:54 +0000
  * doc/index.html: reorder, cleanup links to distributions; add Linux
    from Scratch, Finnix, VServer.
  * sv.c: properly format status command's output on failure cases.
  * sv.c: support optional LSB init script actions reload and
    try-restart.
  * man/sv.8: "sv exit" does not send a TERM signal to the log service
    (thx Jonathan Nieder).
  * fmt_ptime.c: 64 bits time_t fix for svlogd (tnx Jérémie
    Courrèges-Anglas).
  * sv.c: fix typo that may lead to wrong output from sv when reporting
    status of multiple service directories.
  * doc/index.html: add deepOfix Mail Server to list of distributions
    that include runit (thx Debayan Banerjee).
  * doc/useinit.html: update description on how to use with upstart
    (thx Gabriel de Perthuis).
  * doc/index.html: add Dragora GNU/Linux to list of distributions that
    use runit as default init scheme (thx Matias A. Fonzo).
  * doc/index.html: add ArchLinux to list of distributions that include
    runit (thx Kevin Berry).

2.1.1
Sun, 04 Oct 2009 20:28:38 +0000
  * doc/upgrade.html: fix typo.
  * sv.c: on 'down', send runsv the 'down' command properly if not yet
    done (e.g. when taken up with 'once').

2.1.0
Thu, 24 Sep 2009 22:49:33 +0000
  * pathexec_env.c, pathexec.h: add function pathexec_env_run().
  * chpst.c, man/chpst.8: new option -b argv0: run prog with different
    0th argument.
  * sv.c: fix exit code of status command in lsb mode (thx Mathieu
    Poumeyrol)

src/utmpset.c
#include <fcntl.h>
#include <time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include "uw_tmp.h"
#include "strerr.h"
#include "sgetopt.h"
#include "seek.h"
#include "str.h"
#include "open.h"
#include "byte.h"
#include "lock.h"

#define USAGE " [-w] line"
#define FATAL "utmpset: fatal: "
#define WARNING "utmpset: warning: "

const char *progname;

void usage(void) { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }

int utmp_logout(const char *line) {
  int fd;
  unsigned int len;
  uw_tmp ut;
  time_t t;
  int ok =-1;

  if ((len =str_len(line)) > sizeof(ut.ut_line))
    strerr_die4x(111, FATAL,
                 "unable to logout line ", line, " in utmp: name too long");
  if (len < sizeof(ut.ut_line)) ++len;
  if ((fd =open(UW_TMP_UFILE, O_RDWR, 0)) < 0)
    strerr_die4sys(111, FATAL, "unable to open ", UW_TMP_UFILE, ": ");
  if (lock_ex(fd) == -1)
    strerr_die4sys(111, FATAL, "unable to lock: ", UW_TMP_UFILE, ": ");

  while (read(fd, &ut, sizeof(uw_tmp)) == sizeof(uw_tmp)) {
    if (!ut.ut_name[0] || (byte_diff(ut.ut_line, len, line) != 0)) continue;
    memset(ut.ut_name, 0, sizeof ut.ut_name);
    memset(ut.ut_host, 0, sizeof ut.ut_host);
    if (time(&t) == -1) break;
    ut.ut_time = t;
#ifdef DEAD_PROCESS
    ut.ut_type =DEAD_PROCESS;
#endif
    if (lseek(fd, -(off_t)sizeof(uw_tmp), SEEK_CUR) == -1) break;
    if (write(fd, &ut, sizeof(uw_tmp)) != sizeof(uw_tmp)) break;
    ok =1;
    break;
  }
  close(fd);
  return(ok);
}
int wtmp_logout(const char *line) {
  int fd;
  int len;
  struct stat st;
  uw_tmp ut;
  time_t t;

  if ((fd = open_append(UW_TMP_WFILE)) == -1)
    strerr_die4sys(111, FATAL, "unable to open ", UW_TMP_WFILE, ": ");
  if (lock_ex(fd) == -1)
    strerr_die4sys(111, FATAL, "unable to lock ", UW_TMP_WFILE, ": ");

  if (fstat(fd, &st) == -1) {
    close(fd);
    return(-1);
  }
  memset(&ut, 0, sizeof(uw_tmp));
  if ((len =str_len(line)) > sizeof ut.ut_line) len =sizeof ut.ut_line -2;
  byte_copy(ut.ut_line, len, line);
  if (time(&t) == -1) {
    close(fd);
    return(-1);
  }
  ut.ut_time = t;
#ifdef DEAD_PROCESS
  ut.ut_type =DEAD_PROCESS;
#endif
  if (write(fd, &ut, sizeof(uw_tmp)) != sizeof(uw_tmp)) {
    ftruncate(fd, st.st_size);
    close(fd);
    return(-1);
  }
  close(fd);
  return(1);
}

int main (int argc, char * const *argv, const char * const *envp) {
  int opt;
  int wtmp =0;

  progname =*argv;

  while ((opt =getopt(argc, argv, "wV")) != opteof) {
    switch(opt) {
    case 'w':
      wtmp =1;
      break;
    case 'V':
      strerr_warn1("$Id$", 0);
    case '?':
      usage();
    }
  }
  argv +=optind;

  if (! argv || ! *argv) usage();
  if (utmp_logout(*argv) == -1)
    strerr_die4x(111, WARNING, "unable to logout line ", *argv,
                 " in utmp: no such entry");
  if (wtmp)
    if (wtmp_logout(*argv) == -1)
      strerr_die4sys(111, WARNING,
                     "unable to logout line ", *argv, " in wtmp: ");
  _exit(0);
}
